### PR TITLE
 Add artifact cache to track artifacts for faster restart

### PR DESCRIPTION
**Pull request by priyawadhwa**
I added an artifact cache (stored in ~/.skaffold/cache) which maps a
hash of the artifact's dependencies : image name and image tag. I used
kaniko's snapshotting library, which does exactly this for kaniko
caching.

Before building an artifact, we check the cache to see if an image with
the same hash and image name has already been built. If it has, we skip
rebuilding this image. If either the hash or the image name is
different, skaffold will rebuild the image.

The cache can be disabled with the --cache-artifacts=false flag.

Future work could include optimization around retagging cached images
(if we find the hash for an image, but the image name is different,
perhaps we can try to retag the image instead of rebuilding). 

I also wasn't sure if we need to validate that an image exists in a remote repository before deciding to skip the build (how often do people delete their remote images?)

ref #1517

### CODE REVIEW

File: pkg/skaffold/build/cache.go [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+	}
+	return ioutil.WriteFile(c.cacheFile, data, 0755)
+}
+
+func getHashForArtifact(ctx context.Context, a *latest.Artifact) (string, error) {
+	deps, err := DependenciesForArtifact(ctx, a)
+	if err != nil {
+		return "", errors.Wrapf(err, "getting dependencies for %s", a.ImageName)
+	}
+	lm := snapshot.NewLayeredMap(kanikoutil.Hasher(), kanikoutil.CacheHasher())
```
Comment (dgageot) [inline pkg/skaffold/build/cache.go]: Can't we use just a plain digest here?

File: docs/content/en/docs/references/cli/_index.md [hunk 2]
```diff
@@ -74,12 +75,14 @@ Flags:
       --toot                         Emit a terminal beep after the deploy is complete
 
 Global Flags:
-  -v, --verbosity string   Log level (debug, info, warn, error, fatal, panic) (default "warning")
+      --azure-container-registry-config string   Path to the file containing Azure container registry configuration information.
```
[the remark this answers is not in the document]
Comment (priyawadhwa) [reply docs/content/en/docs/references/cli/_index.md]: I think that's coming in from a dependency somehow, let me try to fix it

File: pkg/skaffold/build/cache.go (hunk 1 above)
Comment (priyawadhwa) [reply pkg/skaffold/build/cache.go]: Sorry, what digest do you mean?

**Comment by priyawadhwa** [comment]
@dgageot great points!

> we shouldn't cache the tags and image names. We should be able to realize that an image with the same content was already built, maybe with a different name and tag

I was thinking in another PR we could implement "retagging", so as long as we know the name of one image that has been built we can rename it to whatever we want. 

>double check images by imageId or digest because our local cache might be outdate. Specially locally because I can reset the local Docker very easily.

Sorry if this is a simple question, but do the images in local Docker matter? I assumed that if an image existed in `.skaffold/cache` it should also exist in a remote registry, so upon deploy the cluster should be able to pull it down. 

I like the idea of tagging by content digest -- then we would just need to check that the image exists remotely. Would this interfere with the tagging refactoring you've been doing?

File: docs/content/en/docs/references/cli/_index.md [hunk 3]
```diff
@@ -98,7 +101,8 @@ Usage:
   skaffold completion bash [flags]
 
 Global Flags:
-  -v, --verbosity string   Log level (debug, info, warn, error, fatal, panic) (default "warning")
+      --azure-container-registry-config string   Path to the file containing Azure container registry configuration information.
```
[the remark this answers is not in the document]
Comment (priyawadhwa) [reply docs/content/en/docs/references/cli/_index.md]: fixed

**Comment by priyawadhwa** [comment]
I made significant changes to this on another branch, closing in favor of that one.

**Pull request closed without merging**